Unit Test

german-e-mas opened the pull request:
In this PR I include the first unit tests.

The class testAircraft includes tests for Movement, Crashing and Landing.

I tryed to verify each test from different angles and follow the advices of the "Pragmatic Unit Testing in Java with JUnit" book.

ptal @pablisho

pablisho commented on app/src/test/java/mas/german/landingplanes/TestAircraft.java:
```diff
@@ -0,0 +1,144 @@
+/**
+ * Unit Tests related to Aircraft.
+ */
+public class TestAircraft extends TestCase {
```
I think the @ Test annotation is better than extending TestCase

pablisho commented on app/src/test/java/mas/german/landingplanes/TestAircraft.java:
```diff
@@ -0,0 +1,144 @@
+     * correct angle, according to their direction.
+     */
+    @Test
+    public void testLanding() {
```
It's not necessary to have only one test method per class method. You could split this big test is some smaller ones, so in case it's failing you have a better insight of what's failing of the method. For example in one test you could test the types, in other one the positions, and last the angles.

pablisho commented on app/src/main/java/mas/german/landingplanes/landingsites/LandingSite.java:
```diff
@@ -37,10 +37,16 @@ public Position getPosition() {
         if (direction > Math.PI) {
             direction -= 2 * Math.PI;
         }
+        if (upperLimit > Math.PI) {
```
Could you move this closer to where you calculate upperLimit and lowerLimit?

pablisho commented on app/src/test/java/mas/german/landingplanes/TestLandingSite.java:
```diff
@@ -0,0 +1,88 @@
+
+        // Center and Aperture angles range from 0° to 360°.
+        // Verify all the possible center and aperture angles. In each step they change 10°.
+        for (center = 0; center < 2 * Math.PI; center += Math.toRadians(10)) {
```
It's ok to test all these directions but what it's more usual is to explicitly test the border values, some value outside and some value inside.

pablisho commented on app/src/test/java/mas/german/landingplanes/TestLandingSite.java:
```diff
@@ -0,0 +1,88 @@
+                    }
+                    if (lowerLimitAngle > Math.PI) {
+                        lowerLimitAngle -= 2 * Math.PI;
+                    }
```
The problem here is that if you are using the same code to calculate these limits as the one used in the function you may be not testing anything. If you have some fixed values that you know if should pass and some other you know it shouldn't it may be better.

german-e-mas commented:
I addressed the comments. The test now tries two cases that proved to bring issues, which is when the aperture range contained 0° or 180°, and another normal case. The verification algorithm of the Model was changed, as it wasn't working well in these cases either. Everything works now.

Do you want to take another look @pablisho ?

pablisho commented on app/src/main/java/mas/german/landingplanes/landingsites/LandingSite.java:
```diff
@@ -35,14 +35,18 @@ public Position getPosition() {
+        // direction.
+        // For ease of calculation in the verification, map the angles from [0, 2*PI) to [-PI, PI)
+        // and calculate their difference, which is mapped as well.
+        double angleDifference = mapAngle(direction) - mapAngle(mCenterAngle);
```
For me it's not clear why it's necessary to convert them to [-Pi;PI]

german-e-mas replied:
I think this would be better to explain in our sync later today, but I agree with your point of view!

german-e-mas commented:
I fixed the algorithm based on what we spoke yesterday. It's way more simple now and works like a charm. I will merge this now. Thanks for the reviews and the comments @pablisho !
